### PR TITLE
[Merged by Bors] - chore: generalize more `DirectSum` results to avoid negation

### PR DESCRIPTION
**Pull request by bustercopley**
Generalize `Basis.tensorProduct` and `instance Module.Free.tensor` to CommSemiring and AddCommMonoid, as a follow-up to #10828.

Plus style fixes:
* tidy up LinearAlgebra.DirectSum.Finsupp.lean:
  * delete unused `universe` declaration
  * delete duplicate `open TensorProduct`
  * delete redundant `noncomputable` annotation and associated comment (due to `noncomputable section` in the same file)
  * use `variable`
* delete "typeclass reminders" in "Module.Flat.iff_rTensor_injective'" (no longer needed)
* whitespace fixes
* simplify file structure (`section`/`namespace`/`open`/`variable`)

---

I started by pulling on a thread (deleting `open scoped Classical` from `finsuppTensorFinsupp`), but that deletion itself will be in another PR (#10974).

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by eric-wieser** [comment]
> replace `open Classical` with `DecidableEq` type class implicit arguments
> * add `DecidableEq` elsewhere when needed as a result

This is contentious enough that I'd prefer to see it in its own PR, and not mixed with the rest of this one. Otherwise these changes look good, thanks!

**Comment by eric-wieser** [comment]
bors d+

Thanks!

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean [hunk 1]
```diff
@@ -16,43 +16,30 @@ The tensor product of `ι →₀ M` and `κ →₀ N` is linearly equivalent to
...
+open DirectSum Set LinearMap Submodule TensorProduct
 
-open TensorProduct
-
-open TensorProduct
+variable (R M N ι κ : Type*) [CommSemiring R] [AddCommMonoid M] [Module R M]
+  [AddCommMonoid N] [Module R N]
 
 open scoped Classical in
--- `noncomputable` is a performance workaround for mathlib4#7103
```
[the remark this answers is not in the document]
Comment (bustercopley) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:34]: I'm clearly misunderstanding something! I don't know what effect `noncomputable` has inside a `noncomputable section`. I assumed it had none.

We could also benchmark with `suppress_compilation` ... I'll try that.

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:34]: I think the reality is that `suppress_compilation` in a super early file has a downstream effect, so none of these change make any performance difference.

[the remark this answers is not in the document]
Comment (eric-wieser) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:34]: Let's just reverse the last commit to keep the diff slightly smaller and merge it.

**Comment by bustercopley** [comment]
!bench

File: Mathlib/LinearAlgebra/DirectSum/Finsupp.lean [hunk 2]
```diff
@@ -16,43 +16,31 @@ The tensor product of `ι →₀ M` and `κ →₀ N` is linearly equivalent to
...
-open Set LinearMap Submodule
-
-section TensorProduct
+open DirectSum Set LinearMap Submodule TensorProduct
 
-open TensorProduct
-
-open TensorProduct
+variable (R M N ι κ : Type*) [CommSemiring R] [AddCommMonoid M] [Module R M]
+  [AddCommMonoid N] [Module R N]
```
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:25]: A style nit: putting the types on a separate line to the typeclasses makes adding new types to the section slightly less painful, as then you don't have to rewrap everything.

Comment (bustercopley) [reply Mathlib/LinearAlgebra/DirectSum/Finsupp.lean:25]: Will do

**Pull request closed without merging**